crates/bevy_render/src/mesh/shape/cone.rs

```diff
@@ -0,0 +1,139 @@
+    /// above, this is the number of points you will see on the circle.
+    /// A higher number will make it appear more circular.
+    pub resolution: u32,
+    /// The number of segments between the two ends. Setting this to 1 will have triangles spanning the full
+    /// height of the cone. Setting it to 2 will have two sets of triangles with a horizontal slice in the middle of
+    /// the cone. Greater numbers increase triangles/slices in the same way.
+    pub segments: u32,
+}
+
+impl Default for Cone {
+    fn default() -> Self {
+        Self {
+            top_radius: 0.2,
+            bottom_radius: 0.8,
+            height: 1.0,
+            resolution: 16,
+            segments: 1,
+        }
+    }
+}
+
+impl From<Cone> for Mesh {
+    fn from(c: Cone) -> Self {
+        debug_assert!(c.top_radius >= 0.0);
+        debug_assert!(c.bottom_radius >= 0.0);
+        debug_assert!(c.height > 0.0);
+        debug_assert!(c.resolution > 2);
+        debug_assert!(c.segments > 0);
+
+        let num_rings = c.segments + 1;
+        let num_vertices = c.resolution * 2 + num_rings * (c.resolution + 1);
+        let num_faces = c.resolution * (num_rings - 2);
+        let num_indices = (2 * num_faces + 2 * (c.resolution - 1) * 2) * 3;
+
+        let mut positions = Vec::with_capacity(num_vertices as usize);
+        let mut normals = Vec::with_capacity(num_vertices as usize);
+        let mut uvs = Vec::with_capacity(num_vertices as usize);
+        let mut indices = Vec::with_capacity(num_indices as usize);
+
+        let step_theta = std::f32::consts::TAU / c.resolution as f32;
+        let step_y = c.height / c.segments as f32;
+        let step_radius = (c.top_radius - c.bottom_radius) / c.segments as f32;
+
+        // rings
+
+        for ring in 0..num_rings {
+            let y = -c.height / 2.0 + ring as f32 * step_y;
+            let radius = c.bottom_radius + ring as f32 * step_radius;
+
+            for segment in 0..=c.resolution {
+                let theta = segment as f32 * step_theta;
+                let (sin, cos) = theta.sin_cos();
+
+                positions.push([radius * cos, y, radius * sin]);
+                normals.push([cos, 0., sin]);
+                uvs.push([
+                    segment as f32 / c.resolution as f32,
+                    ring as f32 / c.segments as f32,
+                ]);
+            }
+        }
+
+        // barrel skin
+
+        for i in 0..c.segments {
+            let ring = i * (c.resolution + 1);
+            let next_ring = (i + 1) * (c.resolution + 1);
+
+            for j in 0..c.resolution {
+                indices.extend_from_slice(&[
+                    ring + j,
+                    next_ring + j,
+                    ring + j + 1,
+                    next_ring + j,
+                    next_ring + j + 1,
+                    ring + j + 1,
+                ]);
+            }
+        }
+
+        // caps
+
+        let mut build_cap = |top: bool| {
```

Comment:
How does this behave when top or bottom radius are 0? Would this still generate zero-length edges?

Reply:
Added an explicit check for this now.
If a radius is 0 the generated mesh now draws `138` vertices down from `180`.